File: src/Wc3FrameExtLua/Origin/ChatBox.lua
--=========
-- Include
--=========

---@type FrameExtClass
local FrameExt = require('FrameExt')

--========
-- Module
--========

if not IsGame() then
    return
end

local handle = BlzGetOriginFrame(ORIGIN_FRAME_CHAT_MSG, 0)
BlzFrameClearAllPoints(handle)

---@class FrameExtOriginChatBox
local ChatBox = FrameExt.link(handle, false)
ChatBox:setParent(nil)

return ChatBox

File: src/FrameExt/Origin/ChatBox.ts
import * as Utils from '../../Utils'

import { onPreInit } from '../Init'
import { Frame } from '../Frame'

export class OriginChatBox extends Frame {
    public static inst(){
        if (!OriginChatBox.__instance){
            return error(OriginChatBox.name +
                        ': can not get origin frame before FrameExt finish initialization.', 2)
        }
        return OriginChatBox.__instance
    }

    private constructor(handle: jframehandle){
        super(handle, false)
    }

    private static __instance: OriginChatBox = <OriginChatBox><unknown>undefined
    private static __pre_init_action = (()=>{
        return onPreInit(()=>{
            let handle = BlzGetOriginFrame(ORIGIN_FRAME_CHAT_MSG, 0)
            if (!handle){
                error(OriginChatBox.name + ': failed to init.')
            }

            BlzFrameClearAllPoints(handle)
            BlzFrameSetParent(handle, BlzGetFrameByName("ConsoleUIBackdrop", 0))
        
            OriginChatBox.__instance = new OriginChatBox(handle)
        })
    })()
}